src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManagerImpl.java

```diff
@@ -0,0 +1,227 @@
+    private int currentDepth = -1;
+
+    public boolean isOpen() {
+        return currentDepth > -1;
+    }
+
+    public Transaction openOuter() {
+        if (isOpen()) {
+            throw new IllegalStateException("An outer transaction is already active on this thread.");
+        }
+
+        return open();
+    }
+
+    @Nullable
+    public TransactionContext getCurrentUnsafe() {
+        if (currentDepth == -1) {
+            return null;
+        } else if (stack.get(currentDepth).lifecycle == Transaction.Lifecycle.OPEN) {
+            return stack.get(currentDepth);
+        } else {
+            throw new IllegalStateException("May not call getCurrentUnsafe() from a close callback.");
+        }
+    }
+
+    /**
+     * Open a new transaction, outer or nested, without performing any state check.
+     */
+    Transaction open() {
+        currentDepth++;
+
+        if (stack.size() == currentDepth) {
+            stack.add(new TransactionImpl(currentDepth));
+        }
+
+        TransactionImpl current = stack.get(currentDepth);
+        current.lifecycle = Transaction.Lifecycle.OPEN;
+        return current;
+    }
+
+    void validateCurrentThread() {
+        if (Thread.currentThread() != thread) {
+            String errorMessage = String.format(
+                    "Attempted to access transaction state from thread %s, but this transaction is only valid on thread %s.",
+                    Thread.currentThread().getName(),
+                    thread.getName());
+            throw new IllegalStateException(errorMessage);
+        }
+    }
+
+    public Transaction.Lifecycle getLifecycle() {
+        if (currentDepth == -1) {
+            return Transaction.Lifecycle.NONE;
+        } else {
+            return stack.get(currentDepth).lifecycle;
+        }
+    }
+
+    private class TransactionImpl implements Transaction {
+        final int nestingDepth;
+        final ArrayList<CloseCallback> closeCallbacks = new ArrayList<>();
+        Lifecycle lifecycle = Lifecycle.NONE;
+
+        TransactionImpl(int nestingDepth) {
+            this.nestingDepth = nestingDepth;
+        }
+
+        void validateCurrentTransaction() {
+            validateCurrentThread();
+
+            if (currentDepth == -1 || stack.get(currentDepth) != this) {
+                String errorMessage = String.format(
+                        "Transaction function was called on a transaction with depth %d, but the current transaction has depth %d.",
+                        nestingDepth,
+                        currentDepth);
+                throw new IllegalStateException(errorMessage);
+            }
+        }
+
+        // Validate that this transaction is open.
+        private void validateOpen() {
+            if (lifecycle != Lifecycle.OPEN) {
+                throw new IllegalStateException("Transaction operation cannot be applied to a closed transaction.");
```

Comment:
While I suspect the implementation of both is technically correct, it definitely feels a bit confusing having this `validateOpen` method ensure it is exactly open, but having `Transaction#isOpen` return true when it is open or is still in the process of closing.